snappy-sys/build.rs

```diff
@@ -3,35 +3,99 @@ extern crate cmake;
-    println!("cargo:rustc-link-search=native={}", build.display());
-    fs::copy(src.join("snappy.h"), build.join("snappy.h")).unwrap();
-    if cfg!(target_os = "windows") {
-        fs::copy(stub, build.join("snappy-stubs-public.h")).unwrap();
+    let want_static = env::var("SNAPPY_SYS_STATIC").unwrap_or(String::new()) == "1";
+    let from_source = env::var("SNAPPY_SYS_STATIC_FROM_SOURCE").unwrap_or(String::new()) == "1";
+    if !from_source && configure_snappy(want_static) {
+        return;
+    }
+    build_snappy();
+}
+
+fn configure_snappy(want_static: bool) -> bool {
+    // ~ try pkg_config first
+    if pkg_config::probe_library("snappy").is_ok() {
+        return true;
+    }
+    // ~ then try search in statically predefined directories
+    let libsnappy_file = if want_static { "libsnappy.a" } else { "libsnappy.so" };
+    if let Some(path) = first_path_with_file(libsnappy_file) {
+        if want_static {
+            println!("cargo:rustc-link-search={}", path);
+            println!("cargo:rustc-link-lib=static=snappy");
+            configure_stdcpp();
+        } else {
+            println!("cargo:rustc-link-search=native={}", path);
+            println!("cargo:rustc-link-lib=dylib=snappy");
+        }
+        return true;
+    }
+    return false;
+}
+
+fn build_snappy() {
+    let out_dir = PathBuf::from(&env::var("OUT_DIR").unwrap());
+
+    let cc = cc::Build::new().cpp(true).get_compiler();
+    let mut cflags = OsString::new();
+    for arg in cc.args() {
+        cflags.push(arg);
+        cflags.push(" ");
     }
+    let output = cmake::Config::new("snappy")
+        .env("CC", cc.path())
+        .env("CFLAGS", cflags)
+        .static_crt(true)
+        .build();
+    Command::new("make").current_dir(&output).status().unwrap();
+    println!("cargo:rustc-link-lib=static=snappy");
+    // On some machines, we see lib, on other machines, we see lib64. Add both:
+    println!("cargo:rustc-link-search=native={}/lib", output.display());
+    println!("cargo:rustc-link-search=native={}/lib64", output.display());
+    println!("cargo:root={}", out_dir.to_string_lossy());
     configure_stdcpp();
 }
 
 fn configure_stdcpp() {
-    // From: https://github.com/alexcrichton/gcc-rs/blob/master/src/lib.rs
+    // From: https://github.com/alexcrichton/cc-rs/blob/master/src/lib.rs
     let target = env::var("TARGET").unwrap();
-    let cpp = if target.contains("darwin") { "c++" } else { "stdc++" };
-    println!("cargo:rustc-link-lib={}", cpp);
+    let cpp = if target.contains("darwin") {
+        Some("c++")
+    } else if target.contains("windows") {
+        None
+    } else {
+        Some("stdc++")
+    };
+    if let Some(cpp) = cpp {
+        println!("cargo:rustc-link-lib={}", cpp);
+    }
 }
 
+fn first_path_with_file(file: &str) -> Option<String> {
+    // we want to look in LD_LIBRARY_PATH and then some default folders
+    if let Some(ld_path) = env::var_os("LD_LIBRARY_PATH") {
```

Comment:
I suggest to invoke CC directly to test whether they are configured properly instead. It makes things so complicated. And on Windows, the environment variables should be `LIBPATH`.

Reply:
Hm, now that I'm looking at your `static-link` branch I think I can just remove this.

Reply:
PTAL again :)